### PR TITLE
[SPARK-47739][SQL] Register logical avro type

### PR DESCRIPTION
**Pull request by milastdbx**
### What changes were proposed in this pull request?
In this pull request I propose that we register logical avro types when we initialize `AvroUtils` and `AvroFileFormat`, otherwise for first schema discovery we might get wrong result on very first execution after spark starts.
image

example
```scala
val new_schema = """
     | {
     |   "type": "record",
     |   "name": "Entry",
     |   "fields": [
     |     {
     |       "name": "rate",
     |       "type": [
     |         "null",
     |         {
     |           "type": "long",
     |           "logicalType": "custom-decimal",
     |           "precision": 38,
     |           "scale": 9
     |         }
     |       ],
     |       "default": null
     |     }
     |   ]
     | }""".stripMargin
spark.read.format("avro").option("avroSchema", new_schema).load().printSchema // maps to long - WRONG
spark.read.format("avro").option("avroSchema", new_schema).load().printSchema // maps to Decimal - CORRECT
``` 

### Why are the changes needed?
To fix issue with resolving avro schema upon spark startup.

### Does this PR introduce _any_ user-facing change?
No, its a bugfix

### How was this patch tested?
Unit tests

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala [hunk 1]
```diff
@@ -82,6 +82,35 @@ abstract class AvroLogicalTypeSuite extends QueryTest with SharedSparkSession {
     }
   }
 
+  test("custom logical type registration test") {
```
Comment (dongjoon-hyun) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala:85]: nit.
```
- test("custom logical type registration test") {
+ test("SPARK-47739: custom logical type registration test") {
```

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala [hunk 2]
```diff
@@ -82,6 +82,35 @@ abstract class AvroLogicalTypeSuite extends QueryTest with SharedSparkSession {
...
+    val avroTypeJson =
+      s"""
```
Comment (dongjoon-hyun) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala:87]: We don't need a string interpolation here, do we?

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala [hunk 3]
```diff
@@ -43,6 +43,9 @@ import org.apache.spark.sql.types._
 import org.apache.spark.util.Utils
 
 private[sql] object AvroUtils extends Logging {
+
+  AvroFileFormat.registerCustomAvroTypes()
```
Comment (dongjoon-hyun) [inline connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroUtils.scala:47]: Could you provide a test coverage for this line? The test case on this PR passes without this.

File: connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala [hunk 4]
```diff
@@ -43,6 +43,8 @@ import org.apache.spark.util.SerializableConfiguration
 private[sql] class AvroFileFormat extends FileFormat
   with DataSourceRegister with Logging with Serializable {
 
+  AvroFileFormat.registerCustomAvroTypes()
```
Comment (dongjoon-hyun) [inline connector/avro/src/main/scala/org/apache/spark/sql/avro/AvroFileFormat.scala:46]: ditto. Could you provide a test coverage for this line?  The test case on this PR passes without this.

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala [hunk 5]
```diff
@@ -82,6 +82,35 @@ abstract class AvroLogicalTypeSuite extends QueryTest with SharedSparkSession {
...
+         |      ],
+         |      "default": null
+         |    }
+         |  ]
+         |}
+         |
+   """.stripMargin
+
+    val df = spark.read.format("avro").option("avroSchema", avroTypeJson).load()
+    assert(df.schema.fields(0).dataType == DecimalType(38, 9))
```
Comment (dongjoon-hyun) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala:111]: According to the PR description, this should fail. However, this succeeds already.

```
$ git diff master
diff --git a/connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala b/connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala
index 0a6ebb9e0a7..f54cb28f6ce 100644
--- a/connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala
+++ b/connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeSuite.scala
@@ -82,6 +82,35 @@ abstract class AvroLogicalTypeSuite extends QueryTest with SharedSparkSession {
     }
   }

+  test("custom logical type registration test") {
+    val avroTypeJson =
+      s"""
+         |{
+         |  "type": "record",
+         |  "name": "Entry",
+         |  "fields": [
+         |    {
+         |      "name": "test_col",
+         |      "type": [
+         |        "null",
+         |        {
+         |          "type": "long",
+         |          "logicalType": "custom-decimal",
+         |          "precision": 38,
+         |          "scale": 9
+         |        }
+         |      ],
+         |      "default": null
+         |    }
+         |  ]
+         |}
+         |
+   """.stripMargin
+
+    val df = spark.read.format("avro").option("avroSchema", avroTypeJson).load()
+    assert(df.schema.fields(0).dataType == DecimalType(38, 9))
+  }
+
   // scalastyle:off line.size.limit
   val timestampSchema = s"""
       {

$ build/sbt "avro/testOnly *Avro*LogicalTypeSuite" -Pavro
...
[info] AvroV1LogicalTypeSuite:
[info] - Logical type: date (2 seconds, 227 milliseconds)
[info] - custom logical type registration test (8 milliseconds)
[info] - Logical type: timestamp_millis (258 milliseconds)
[info] - Logical type: timestamp_micros (228 milliseconds)
[info] - Logical type: local_timestamp_millis (232 milliseconds)
[info] - Logical type: local_timestamp_micros (192 milliseconds)
[info] - Logical type: user specified output schema with different timestamp types (160 milliseconds)
[info] - Logical type: user specified output schema with different timestamp ntz types (203 milliseconds)
[info] - Read Long type as Timestamp (73 milliseconds)
[info] - Read Long type as Timestamp without time zone (54 milliseconds)
[info] - Logical type: user specified read schema (92 milliseconds)
[info] - Logical type: Decimal (579 milliseconds)
[info] - Logical type: write Decimal with BYTES type (169 milliseconds)
[info] - Logical type: Decimal with too large precision (93 milliseconds)
[info] - SPARK-43901: LogicalType: Custom Decimal for Long Type (46 milliseconds)
[info] - SPARK-43901: LogicalType: Decimal for Long Type Exception Cases (328 milliseconds)
[info] AvroV2LogicalTypeSuite:
[info] - Logical type: date (375 milliseconds)
[info] - custom logical type registration test (3 milliseconds)
[info] - Logical type: timestamp_millis (252 milliseconds)
[info] - Logical type: timestamp_micros (247 milliseconds)
[info] - Logical type: local_timestamp_millis (227 milliseconds)
[info] - Logical type: local_timestamp_micros (403 milliseconds)
[info] - Logical type: user specified output schema with different timestamp types (236 milliseconds)
[info] - Logical type: user specified output schema with different timestamp ntz types (158 milliseconds)
[info] - Read Long type as Timestamp (75 milliseconds)
[info] - Read Long type as Timestamp without time zone (112 milliseconds)
[info] - Logical type: user specified read schema (114 milliseconds)
[info] - Logical type: Decimal (298 milliseconds)
[info] - Logical type: write Decimal with BYTES type (220 milliseconds)
[info] - Logical type: Decimal with too large precision (67 milliseconds)
[info] - SPARK-43901: LogicalType: Custom Decimal for Long Type (63 milliseconds)
[info] - SPARK-43901: LogicalType: Decimal for Long Type Exception Cases (395 milliseconds)
[info] Run completed in 9 seconds, 471 milliseconds.
[info] Total number of tests run: 32
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 32, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 26 s, completed Apr 6, 2024, 11:17:15 AM
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Thank you for making a PR, @milastdbx . However, this PR doesn't provide any test coverage for the claim. Could you revise your PR? In general, 

1. According to your PR description, the new test case should fail without your patch.
2. We need a new test coverage which fails if we don't have the new line of `class AvroFileFormat`
3. We need a new test coverage which fails if we don't have the new line of `object AvroUtils`

**Comment by milastdbx** [comment]
@dongjoon-hyun thanks for the review. You are absolutely right. Your tests are passing because you are running them in batch, and this regression only really happens if you call don't initalize `LogicalTypes` prior to resolving schema, hence I would assume that some test registered it and as a side effect you get tests passing.

Given that `LogicalTypes` is an external static class, part of `Apache Avro` I am not sure how to resolve this reliably.
One idea is to separate this into different Suite, but this also does not give complete protection, as multiple suites can be run within same process, hence hitting same issue as if test was part of suite.

Do you have any suggestion how to test this static object initalization when we do not have access to `teardown` api of these classes ?

**Comment by milastdbx** [comment]
@dongjoon-hyun I've isolated test in separate suite to address issue you found. This does not eliminate it if multiple suites are run however.

Please take another look and i'd appreciate your feedback

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeInitSuite.scala [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+/**
+ * Test suite for Avro logical type initialization.
+ * Tests here must run in isolation, otherwise some other test might
+ * initialize variable and make this test flaky
+ */
+abstract class AvroLogicalTypeInitSuite
+    extends QueryTest
+      with SharedSparkSession {
+
+  test("custom logical type registration test") {
```
Comment (dongjoon-hyun) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeInitSuite.scala:34]: It seems that you missed this previous comment. Could you address it?
- https://github.com/apache/spark/pull/45895#discussion_r1554648407

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeInitSuite.scala [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.DecimalType
+
+/**
+ * Test suite for Avro logical type initialization.
+ * Tests here must run in isolation, otherwise some other test might
+ * initialize variable and make this test flaky
+ */
+abstract class AvroLogicalTypeInitSuite
+    extends QueryTest
```
Comment (dongjoon-hyun) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeInitSuite.scala:31]: Indentation? Please refer the following guide line and other classes.
- https://github.com/databricks/scala-style-guide?tab=readme-ov-file#spacing-and-indentation

https://github.com/apache/spark/blob/85c4f053f25a7f20546600e179a7303a4409834f/connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroRowReaderSuite.scala#L40-L42

File: connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeInitSuite.scala [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+         |}
+         |
+   """.stripMargin
+
+    val df = spark.read.format("avro").option("avroSchema", avroTypeJson).load()
+    assert(df.schema.fields(0).dataType == DecimalType(38, 9))
+  }
+}
+
+class AvroV1LogicalTypeInitSuite extends AvroLogicalTypeInitSuite {
```
Comment (dongjoon-hyun) [inline connector/avro/src/test/scala/org/apache/spark/sql/avro/AvroLogicalTypeInitSuite.scala:64]: Thank you for adding this test coverage. I verified this.
```
[info] AvroV1LogicalTypeInitSuite:
[info] - custom logical type registration test *** FAILED *** (1 second, 105 milliseconds)
[info]   LongType did not equal DecimalType(38, 9) (AvroLogicalTypeInitSuite.scala:60)
[info]   Analysis:
[info]   LongType$(precision: -> 38, scale: -> 9)
```

**Review by dongjoon-hyun — COMMENTED** [review]
It looks good to me except two minor comments. Thank you for revising the PR, @milastdbx .

**Comment by dongjoon-hyun** [comment]
To @milastdbx , why did you request a review to me when you didn't address my comment?

Screenshot 2024-04-15 at 17 03 41

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
Please address the remaining requests.

**Comment by milastdbx** [comment]
Sorry for the wrong ping. I didn't check whether push succeeded. It turns out it failed because there was a commit inbetween, and i didn't notice.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for updating, @milastdbx . 😄 
Merged to master for Apache Spark 4.0.0.

**Pull request closed without merging**